benchmark/david_benchmark.py

```diff
@@ -0,0 +1,154 @@
+# © Crown Copyright GCHQ
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Benchmark performance of different coreset algorithms on pixel data from an image.
+
+The benchmarking process follows these steps:
+1. Load an input image and downsample (downsampling reduces the image resolution,
+   i.e., the number of points in the dataset, to reduce the computational load on the
+   machine this script is run on. A downsampling factor of 1 corresponds to no
+   downsampling).
+2. Convert to grayscale and extract pixel locations and values.
+3. Generate coresets with different algorithms.
+4. Plot the original image alongside coresets generated by each algorithm.
+5. Save the resulting plot as an output file.
+
+Each coreset algorithm is timed to measure and report the time taken for each step.
+"""
+
+import os
+import time
+from pathlib import Path
+from typing import Optional
+
+import cv2
+import equinox as eqx
+import jax.numpy as jnp
+import matplotlib.pyplot as plt
+import numpy as np
+from jax import random
+from mnist_benchmark import get_solver_name, initialise_solvers
+
+from coreax import Data
+
+MAX_8BIT = 255
+
+
+def downsample_opencv(image_path: str, downsampling_factor: int) -> np.ndarray:
```

Comment:
I'm getting a duplicated code fragment (19 lines long) warning.

Consider importing `downsample_opencv()` from david_map_reduce_weighted.py instead of copying the function here